### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by tpoisot**
This defines one abstract type (`NeutralLandscapeMaker`), the `PlanarGradient`, `EdgeGradient`, and `NoGradient` types and algorithms, and overloads `Base.rand` and `Random.rand!` as the main method.

### CODE REVIEW

File: src/algorithms/planargradient.jl [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    PlanarGradient(x::T) where {T <: Real} = new(mod(x, 360.0))
+end
+
+PlanarGradient() = PlanarGradient(360rand())
+
+function landscape(alg::PlanarGradient, s::Tuple{IT,IT}=(10,30)) where {T <: Real, IT <: Integer}
+    eastness = sin(deg2rad(alg.direction))
+    southness = -1cos(deg2rad(alg.direction))
+    l = southness .* (1:s[1]) .+ (eastness .* (1:s[2]))'
+    return l
```
Comment (mkborregaard) [inline src/algorithms/planargradient.jl]: I see you moved out the masking - I agree with that. The user should be able to mask manually with a mask function, no need to have that as a keyword arg.

File: src/algorithms/planargradient.jl [hunk 2]
```diff
@@ -0,0 +1,13 @@
+struct PlanarGradient <: NeutralLandscapeMaker
+    direction::Float64
+    PlanarGradient(x::T) where {T <: Real} = new(mod(x, 360.0))
+end
+
+PlanarGradient() = PlanarGradient(360rand())
+
+function landscape(alg::PlanarGradient, s::Tuple{IT,IT}=(10,30)) where {T <: Real, IT <: Integer}
```
Comment (mkborregaard) [inline src/algorithms/planargradient.jl]: Why is the size a tuple? Most of the Array constructors, rand etc take two separate args

File: src/NeutralLandscapes.jl [hunk 3]
```diff
@@ -2,4 +2,10 @@ module NeutralLandscapes
 
 greet() = print("Hello World!")
```
Comment (mkborregaard) [inline src/NeutralLandscapes.jl]: I guess we don't need greet

**Review by mkborregaard — COMMENTED** [review]
Yes, great! This is the right appraoch imho

**Comment by gottacatchenall** [comment]
also the python version scales everything between 0 and 1, not sure if that should be an arg to `landscape()` or a separate call

File: src/algorithms/edgegradient.jl [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+EdgeGradient() = EdgeGradient(360rand())
+
+"""
+    landscape(alg::EdgeGradient, s::Tuple{IT,IT}=(10,30)) where {IT <: Integer}
+
+Creates a landscape of size `s` following the planar gradient model. The
+additional arguments `kw...` are passed to the post-processing function, see the
+documentation of **TODO**. 
+"""
+function landscape(alg::EdgeGradient, s::Tuple{IT,IT}=(10,30); kw...) where {IT <: Integer}
```
Comment (mkborregaard) [inline src/algorithms/edgegradient.jl]: Could we have a function instead:
```julia
function landscape(alg, size; mask = Nothing, kw...) 
  ret = Matrix(size...)
  landscape!(alg, size; mask = mask, kw...)
end

function landscape!(mat, alg, size; mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask!(ret)
  rescale!(ret)
end
```
and then have the dispatching function be `_landscape!`?

Comment (mkborregaard) [reply src/algorithms/edgegradient.jl]: would save a lot of the boilerplate in the python version

Comment (mkborregaard) [reply src/algorithms/edgegradient.jl]: @gottacatchenall ^

Comment (gottacatchenall) [reply src/algorithms/edgegradient.jl]: i like this structure but i wonder if there is a case where a user would not want it rescaled---we could do something similar as mask, ie

```
function landscape!(mat, alg, size; scale = true, mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask!(ret)
  scale && rescale!(ret)
end
```

or something similar, although i'm not sure if there is a need for that use-case

File: src/algorithms/planargradient.jl (hunk 2 above)
Comment (tpoisot) [reply src/algorithms/planargradient.jl]: that's mostly shorthand to say "this should have two values", but that's a good point, I can change that

File: src/NeutralLandscapes.jl (hunk 3 above)
Comment (tpoisot) [reply src/NeutralLandscapes.jl]: forgot to remove it!

File: src/algorithms/edgegradient.jl (hunk 4 above)
Comment (tpoisot) [reply src/algorithms/edgegradient.jl]: the issue with not rescaling is that it *does* still scale, but with network size - this is not ideal, I think

Comment (tpoisot) [reply src/algorithms/edgegradient.jl]: > and then have the dispatching function be `_landscape!`?

This is so obvious... Yes!

File: src/algorithms/planargradient.jl (hunk 1 above)
Comment (tpoisot) [reply src/algorithms/planargradient.jl]: so... the only issue is that the scaling is done *after* the masking, which tends to break the flow of a lot of things

Comment (tpoisot) [reply src/algorithms/planargradient.jl]: my current solution would be to mask if a mask argument is given (or a series of coordinates, maybe), and then *also* give a mask function after that

File: src/algorithms/edgegradient.jl (hunk 4 above)
Comment (mkborregaard) [reply src/algorithms/edgegradient.jl]: https://github.com/EcoJulia/NeutralLandscapes.jl/pull/2

Comment (gottacatchenall) [reply src/algorithms/edgegradient.jl]: > the issue with not rescaling is that it _does_ still scale, but with network size - this is not ideal, I think

hmm i guess i'm thinking if you want different types of scaling (linear scaling on an arbitrary interval,  standardization, normalization, etc), you could use something like this to use an arbitrary scaling function

```
function landscape!(mat, alg, size; scaler = scale!, mask = Nothing, kw...)
  _landscape!(mat!, alg, size; kw...)
  isnothing(mask) || mask(ret)
  scaler(ret)
end
```

although if we want to replicate python functionality before adding new things that is good with me too

File: src/algorithms/planargradient.jl (hunk 1 above)
Comment (mkborregaard) [reply src/algorithms/planargradient.jl]: right, my mistake :+1:

File: src/NeutralLandscapes.jl (hunk 3 above)
Comment (mkborregaard) [reply src/NeutralLandscapes.jl]: I took it out in #2

File: src/algorithms/edgegradient.jl (hunk 4 above)
Comment (mkborregaard) [reply src/algorithms/edgegradient.jl]: I'd think so - fidelity first. Scaling is also quite cheap I'd think

**Comment by mkborregaard** [comment]
Instead of `landscape!` and `landscape` - should it just be `rand` and `rand!`? I quite like
```julia
landscape = rand(PlanarGradient(60.0), 100, 120)
```
for its symmetry with
```julia
using Distributions
normalarray = rand(Normal(20, 3), 100, 120)
```

**Comment by tpoisot** [comment]
> Instead of `landscape!` and `landscape` - should it just be `rand` and `rand!`? I quite like

I was wondering the same thing while I was cooking...  I think it makes sense.

**Comment by tpoisot** [comment]
Let's work from `main`!

**Pull request merged**